### PR TITLE
Add sha1 SparkSQL function

### PR DESCRIPTION
**Pull request by izchen**

### CODE REVIEW

**Review by mbasmanova — COMMENTED** [review]
@izchen Please, add documentation for this new function. See https://facebookincubator.github.io/velox/spark_functions.html

File: velox/functions/sparksql/String.cpp [hunk 1]
```diff
@@ -146,4 +146,13 @@ std::shared_ptr<exec::VectorFunction> makeLength(
   return kLengthFunction;
 }
 
+void encodeDoubleLengthHexString(uint8_t *data, int size) {
+  static char const HEX_CODES[] = "0123456789abcdef";
```
Comment (mbasmanova) [inline velox/functions/sparksql/String.cpp:150]: Please, review [coding style](https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md) and update the code to comply.

HEX_CODES -> kHexCodes

File: velox/functions/sparksql/String.h [hunk 2]
```diff
@@ -70,6 +70,23 @@ std::shared_ptr<exec::VectorFunction> makeLength(
     const std::string& name,
     const std::vector<exec::VectorFunctionArg>& inputArgs);
 
+void encodeDoubleLengthHexString(uint8_t *data, int size);
```
Comment (mbasmanova) [inline velox/functions/sparksql/String.h:73]: Please, document this function.

File: velox/functions/sparksql/String.h [hunk 3]
```diff
@@ -70,6 +70,23 @@ std::shared_ptr<exec::VectorFunction> makeLength(
...
+
+template <typename T>
+struct Sha1HexStringFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE
+  void call(out_type<Varbinary>& result, const arg_type<Varbinary>& input) {
+    static const int SHA1_LENGTH = 20;
```
Comment (mbasmanova) [inline velox/functions/sparksql/String.h:81]: naming: kSha1Length

File: velox/functions/sparksql/tests/StringTest.cpp [hunk 4]
```diff
@@ -141,6 +146,12 @@ TEST_F(StringTest, MD5) {
   EXPECT_EQ(md5("Infinity"), "eb2ac5b04180d8d6011a016aeb8f75b3");
 }
 
+TEST_F(StringTest, SHA1) {
```
Comment (mbasmanova) [inline velox/functions/sparksql/tests/StringTest.cpp:149]: naming: SHA1 -> sha1

File: velox/docs/functions/spark/binary.rst [hunk 5]
```diff
@@ -13,3 +13,7 @@ Binary Functions
 .. spark:function:: xxhash64(x) -> integer
 
     Computes the xxhash64 of x.
+
+.. spark:function:: sha1(x) -> string
```
Comment (mbasmanova) [inline velox/docs/functions/spark/binary.rst:17]: Functions in the documentation should appear in alphabetical order. Please, move accordingly.

Comment (mbasmanova) [inline velox/docs/functions/spark/binary.rst:17]: In Spark, this function appears to allow both VARCHAR and VARBINARY input: https://docs.databricks.com/sql/language-manual/functions/sha1.html

This implementation allows only VARBINARY. Should we add support for VARCHAR as well?

Comment (izchen) [reply velox/docs/functions/spark/binary.rst:17]: Thank you for the review!


In [SparkSQL code](https://github.com/apache/spark/blob/e6a84fef2fa43b880c89f2f54a5fa78033021fb1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala#L172), the input type of the sha1 function expression  is binary. Spark implicitly inserts a cast expression when the sha1 input column is String type. So I think we are fine with just supporting VARBINARY.

```
select sha1(a) from testtb2;

== Physical Plan ==
* Project (3)
+- * ColumnarToRow (2)
   +- Scan parquet default.testtb2 (1)

(1) Scan parquet default.testtb2
Output [1]: [a#0]
ReadSchema: struct<a:string>

(3) Project [codegen id : 1]
Output [1]: [sha1(cast(a#0 as binary)) AS sha1(a)#4]
Input [1]: [a#0]
```

Comment (mbasmanova) [reply velox/docs/functions/spark/binary.rst:13]: @izchen Thank you for explaining.

File: velox/docs/functions/spark/binary.rst [hunk 6]
```diff
@@ -10,6 +10,10 @@ Binary Functions
 
     Computes the md5 of x.
 
+.. spark:function:: sha1(x) -> string
```
Comment (mbasmanova) [inline velox/docs/functions/spark/binary.rst:13]: - string -> varchar
- Let's document that 'x' should be of type varbinary.

File: velox/functions/sparksql/tests/StringTest.cpp [hunk 7]
```diff
@@ -141,6 +146,12 @@ TEST_F(StringTest, MD5) {
   EXPECT_EQ(md5("Infinity"), "eb2ac5b04180d8d6011a016aeb8f75b3");
 }
 
+TEST_F(StringTest, sha1) {
+  EXPECT_EQ(sha1(std::nullopt), std::nullopt);
+  EXPECT_EQ(sha1(""), "da39a3ee5e6b4b0d3255bfef95601890afd80709");
+  EXPECT_EQ(sha1("Spark"), "85f5955f4b27a9a4c2aab6ffe5d7189fc298b92c");
```
Comment (mbasmanova) [inline velox/functions/sparksql/tests/StringTest.cpp:152]: perhaps, add a test that uses a very long string, > 20 characters.

**Review by mbasmanova — COMMENTED** [review]
@izchen Looks great % a few minor comments.

**Review by mbasmanova — APPROVED** [review]
@izchen Looks great. Thank you for addressing the comments.

**Comment by mbasmanova** [comment]
@izchen There are CI failures. Would you take a look?

**Pull request closed without merging**